src/pstore/pstore.c

```diff
@@ -0,0 +1,582 @@
+        {
+#if 1
+            r = copy_file_atomic(ifdpath, ofdpath, 0644, 0, 0, COPY_REPLACE);
+            remove_file = (r == 0);
+#else
+            int ofd;
+
+            ofd = open(ofdpath, O_CREAT|O_TRUNC|O_WRONLY, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH/*0644*/);
+            if (ofd < 0)
+            {
+                log_error_errno(errno, "open(%s): %m\n", ofdpath);
+                r = errno;
+            }
+            else
+            {
+                if (dc->contents != NULL)
+                {
+                    ssize_t wr;
+                    wr = write(ofd, dc->contents, dc->statbuf.st_size);
+                    /* Detect write problem */
+                    if (wr != dc->statbuf.st_size)
+                    {
+                        log_error_errno(errno, "write(%s): %m\n", ofdpath);
+                        r = errno;
+                    }
+                    else
+                        remove_file = 1;
+                }
+            }
+            close(ofd);
+// FIX??? update ofd stat.time to same info as dc.stat.time?
+#endif
+        }
+        else
+        {
+            log_error_errno(errno, "mkdir_parents(): %m\n");
+            r = errno;
+        }
+    }
+    else
+    if (arg_storage == PSTORE_STORAGE_JOURNAL)
+    {
+        //sd_journal_send("MESSAGE=%s", dc->de.d_name, NULL);
+        if (dc->is_binary)
+        {
+            //WITH_BINARY= but how to know end of binary blob if contained in 0 terminated string?
+        }
+        else
+        {
+            r = sd_journal_send("MESSAGE=File %s:\n%s", dc->de.d_name, dc->contents, NULL);
+        }
+        /* NOTE: If journald not running, r always 0 */
+        remove_file = (r == 0);
+    }
+
+    /* If file copied properly, remove it from pstore */
+    if ((r == 0) && remove_file && arg_unlink)
+    {
+        unlink(ifdpath);
+    }
+
+    return r;
+}
+
+static void write_dmesg (const char *dmesg, ssize_t size, const char *id) {
+    if ((dmesg != NULL) && (size > 0))
+    {
+        _cleanup_free_ char *ofdpath;
+        int ofd;
+
+        log_info("Record ID %s\n", id);
+
+        ofdpath = path_join(arg_archivedir, id, "dmesg.txt");
+
+        ofd = open(ofdpath, O_CREAT|O_TRUNC|O_WRONLY, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH/*0644*/);
+        if (ofd < 0)
+        {
+            log_error_errno(errno, "open(%s): %m\n", ofdpath);
+        }
+        else
+        {
+            ssize_t wr;
+            wr = write(ofd, dmesg, size);
+            if (wr != size)
+            {
+                log_error_errno(errno, "write(%s): %m\n", ofdpath);
+            }
+        }
+        close(ofd);
+    }
+}
+
+/*
+ * PStore content rule handling
+ */
+typedef struct rules_s
+{
+    void (*handler)(struct rules_s *rule, dirent_ct *dc);
+    const char *handlerName;
+    const char *pattern;
+} rules_t;
+static void dmesg_handler (struct rules_s *rule, dirent_ct *dc);
+static void move_handler (struct rules_s *rule, dirent_ct *dc);
+rules_t rules[] =
```

Comment:
this looks like something that should be `static const`, no?

Reply:
done